src/TensorFlowNET.Core/Buffers/Buffer.cs

```diff
@@ -14,45 +14,44 @@
    limitations under the License.
 ******************************************************************************/
 
+using NumSharp.Backends.Unmanaged;
 using System;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
-using NumSharp.Backends.Unmanaged;
+using Tensorflow.Util;
 using static Tensorflow.c_api;
 
 namespace Tensorflow
 {
     /// <summary>
     ///     Represents a TF_Buffer that can be passed to Tensorflow.
     /// </summary>
-    public class Buffer : DisposableObject
+    public sealed class Buffer : IDisposable
     {
-        private unsafe TF_Buffer buffer
-        {
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => *bufferptr;
-        }
+        public SafeBufferHandle Handle { get; }
 
-        private unsafe TF_Buffer* bufferptr
-        {
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => (TF_Buffer*) _handle;
-        }
+        /// <remarks>
+        /// <para>The result of this method is only valid in the context of a lease on <see cref="Handle"/>. See
+        /// <see cref="SafeHandleExtensions.Lease(SafeHandle)"/>.</para>
```

Comment:
💭 I'm still thinking about ways to describe this. There are actually two valid ways to use it:

1. Direct ownership

    ```csharp
    using (var buffer = new Buffer())
    {
      DoSomethingWithMemory(buffer.DangerousBuffer);
    }
    ```

2. Via a lease on the handle


    ```csharp
    using (var bufferLease = buffer.Handle.Lease())
    {
      DoSomethingWithMemory(buffer.DangerousBuffer);
    }
    ```

Reply:
@Oceania2018 How do you feel about this wording?

> The result of this method is only valid when the underlying `Handle` has not been disposed. If the lifetime of the object is unclear, a lease may be used to prevent disposal while the object is in use. See `SafeHandleExtensions.Lease(SafeHandle)`.

Reply:
Actually, I don't very understand why we need `buffer.Handle.Lease()` for `public ulong Length` ? Do we really need it once there is `SafeBufferHandle`, GC won't collect `Buffer` if there is any reference to it, right?

Reply:
Yes, it's necessary.

1. The implementation of `Length` could get inlined into the caller
2. The call to `DangerousBuffer` is the last true use of the `Buffer` type

The problematic sequence would look like this. I'm using manual inlining to show what could happen.

```
var buffer = new Buffer(); // Note no 'using' statement
var tfBuffer = buffer.DangerousBuffer;
// GC could collect Buffer.Handle here, making tfBuffer invalid
var length = tfBuffer.Length; // Crash!
```

Reply:
📝 Note that `Lease()` does not incur any allocations, and is generally very fast.

Reply:
I havn't test the performance of `Lease()`, because we donsn't use it very much at this moment. Your comment is making sense.